### DISCUSSION TITLE
When registering blocks via register_block_type('block.json'), the asset path is wrong

### DISCUSSION THREAD

**Discussion by jakeparis** [WordPress/gutenberg#40741] [Q&A]
I am building a theme and learning about the concept of registering blocks with `register_block_type('block.json')`. However, when I do, the script/style file asset paths are messed up. Any insights/pointers would be most appreciated as I learn this new way.

The editor seems to be trying to load a url based on 
`https://example.com/mysite/wp-content/plugins/`
concatenated with the filesystem path to the block index file:
`/shared/httpd/htdocs/wp-content/themes/jmc/build/blocks/jake/index.js?ver=192383983`

So the end results look like:
* `https://example.com/mysite/wp-content/plugins/shared/httpd/htdocs/wp-content/themes/jmc/build/blocks/jake/index.js?ver=c10b6a3ac45e2d11a234a6cd2c193a5c`
* `https://example.com/mysite/wp-content/plugins/shared/httpd/htdocs/wp-content/themes/jmc/build/blocks/jake/index.css?ver=5.9.3`

Here's my block.json.

**block.json**
```
{
  ...
	"editorScript": "file:./index.js",
	"editorStyle": "file:./index.css",
	"viewScript": "file:./public.js",
	"style": "file:./public.css"
 }
```

I'm loading all the block.json files in a loop like so:

**registration.php**
```php
foreach ( glob( get_stylesheet_directory() . "/build/blocks/*/block.json" ) as $file ) {
    register_block_type( $file );
}
```

My `src` directory is structured like so:
```
themedir
| - src
| - | - blocks
| - | - | - jake
| - | - | - | - block.json
| - | - | - | - index.js
| - | - | - | - index.scss
                 ...
| - build
| - | - blocks
| - | - | - jake
| - | - | - | - block.json
| - | - | - | - index.js
| - | - | - | - index.css
                 ...
```

**Comment by jakeparis** [comment]
This may be a problem because I am using this technique in a theme, not a plugin. I've filed a ticket: https://github.com/WordPress/gutenberg/issues/40744

**Reply by gziolo** [reply]
This is going to be possible to register blocks a the theme in a way you presented as of WordPress 6.0. See related commit https://core.trac.wordpress.org/changeset/53091 and PR https://github.com/WordPress/wordpress-develop/pull/2494 for full context. @fabiankaegy might provide you an example of a useful filter you can use if you need it earlier.

**Reply by jakeparis** [reply]
Thanks @gziolo . I've downloaded WP 6.0 RC1, and this does work. However, it _only_ works if I don't have an entry path specified in the build command. So for example:

```
wp-scripts build
```

results in something like:
```
build
| - blocks
| - | - block1
| - | - | - block.json
| - | - | - editor.css
| - | - | - public.css
| - | - | - index.js
| - | - | - ...
```
which is good. However, if I use:
```
wp-scripts build theme=./src/theme/index.js
```

Then the data from the block.json's `editorScript`, `editorStyle`, etc fields are _not_ used to create the assets. In this case, only the `block.json` is transferred to the build directory.

The ultimate goal here is a theme build process which can accomodate building blocks like in the first example above, and also build global theme stylesheets/scripts. Is there perhaps another way to build a global theme stylesheet? Not all of my styles are block-specific. In the past, I've used a separate npm command for that: 
```json
{
  "scripts":
    "build:blocks": "wp-scripts build",
    "build:theme": "wp-scripts theme=./src/theme/index.js"      
}
```
but I would guess there's a better way to do this all at once.

**Reply by gziolo** [reply]
Interesting use case. The scripts won’t detect `block.json` files when you explicitly pass entry points. At the moment, you would have to run two scripts as in the example you shared.

We definitely can think how to improve the experience. @ryanwelcher, any thoughts on that? Have you run into something similar in you endeavors? 😃

**Reply by jakeparis** [reply]
The other issue here is that if I use two different build scripts, I believe they will override each other's output in the build directory. 

I'm surprised that no one else has need of creating a theme global stylesheet? Am I thinking about this all wrong in this new world of block themes?

**Reply by jakeparis** [reply]
Nevermind, I solved the "override each other's output" issue by specifying `--output-dir=` in the build command.

**Reply by gziolo** [reply]
Nice, I'm glad you sorted it out. We will keep your use case in mind. It would be cool if you could have everything under control with a single commend and without a custom Webpack config file (which would be another option that other projects often use).

**Reply by kfathy194** [reply]
@jakeparis Hi
I have the exact same scenario as yours, can you please share the end solution?

**Reply by kmorepubli** [reply]
i have it here also

**Reply by gziolo** [reply]
What WordPress version and which platform do you use (Windows, macOS, linux)?
